---
title: "03-hisat2-summer2021_2022"
output: html_document
date: "2024-05-21"
---
Rmd to run `HISAT2` with _Pycnopodia helianthoides_ genome and _Pycnpoodia helianthoides_ coelomocyte RNAseq sequences. 

All trimmed reads from summer 2021 and 2022 live on Raven:       
Summer 2021     
`/home/shared/8TB_HDD_02/graceac9/data/pycno2021`

Summer 2022    
`/home/shared/8TB_HDD_02/graceac9/data/pycno2022` 

Get genome from NCBI into data directory on Raven:        
https://www.ncbi.nlm.nih.gov/datasets/genome/GCA_032158295.1/       

```{r, engine='bash'}
cd ../data

/home/shared/datasets download genome accession GCA_032158295.1 --include gff3,rna,cds,protein,genome,seq-report
```

```{r, engine='bash'}
cd ../data
unzip ncbi_dataset.zip

```


```{r, engine='bash', eval=TRUE}
ls ../data/ncbi_dataset/data/

```

```{r, engine='bash', eval=TRUE}
ls ../data/ncbi_dataset/data/GCA_032158295.1

```

Get gtf file on raven into data folder:    
I downloaded it originally from 

rsync /Users/graciecrandall/Downloads/augustus.hints.gtf graceac9@raven.fish.washington.edu:/home/shared/8TB_HDD_02/graceac9/GitHub/paper-pycno-sswd-2021-2022/data
graceac9@raven.fish.washington.edu's password: 

```{r, engine='bash', eval=TRUE}
head ../data/augustus.hints.gtf

```

For `stringtie` later, the gtf needs to be in gff format. There isn't currently one in existence that works. Steven figured out how to fix this issue (https://sr320.github.io/tumbling-oysters/posts/sr320-20-seastar/)  

Won't run code because Steven did this already and the new file lives in:
` ../analyses/12-fix-gff/mod_augustus.gtf`


```{r}
library(knitr)
library(tidyverse)
library(Biostrings)
library(pheatmap)
library(DESeq2)
library(tidyverse)
```


# 1. BUILD AN INDEX

Follow this code: https://htmlpreview.github.io/?https://github.com/urol-e5/deep-dive/blob/8fd4ad4546d1d95464952f0509406efd9e42ffa0/D-Apul/code/04-Apulcra-hisat.html 

```{bash}
pwd
```

From the gtf, get exon list
```{bash}
/home/shared/hisat2-2.2.1/hisat2_extract_exons.py \
../analyses/12-fix-gff/mod_augustus.gtf \
> ../analyses/03-hisat2/m_exon.tab
```

```{bash}
head ../analyses/03-hisat2/m_exon.tab
```

from the gtf, get the splice sites
```{bash}
#!/bin/bash

# This script will extract splice sites from the gtf file

# This is the command to extract splice sites from the gtf file
/home/shared/hisat2-2.2.1/hisat2_extract_splice_sites.py \
../analyses/12-fix-gff/mod_augustus.gtf \
> ../analyses/03-hisat2/m_splice_sites.tab

```

use the genome fasta to make an index for alignment
```{bash}
# build an index 
/home/shared/hisat2-2.2.1/hisat2-build \
../data/ncbi_dataset/data/GCA_032158295.1/GCA_032158295.1_ASM3215829v1_genomic.fna \
../data/Phel_genome_index \
--exon ../analyses/03-hisat2/m_exon.tab \
--ss ../analyses/03-hisat2/m_splice_sites.tab \
-p 40 \
../analyses/12-fix-gff/mod_augustus.gtf \
2> ../analyses/03-hisat2/hisat2-build_stats.txt
```

```{bash}
tail ../analyses/03-hisat2/hisat2-build_stats.txt
```

# 2. ALIGNMENT


```{bash}
for file in ../data/2021_trimmed/*_R1*fq.gz; do
    # Remove the  part to get the base name
    base=$(basename "$file" _R1_001.fastq.gz.fastp-trim.20220810.fq.gz)
    # Construct the names of the pair of files
    file1=${base}_R1_001.fastq.gz.fastp-trim.20220810.fq.gz
    file2=${base}_R2_001.fastq.gz.fastp-trim.20220810.fq.gz
    # Run the hisat2 command
    /home/shared/hisat2-2.2.1/hisat2 \
    -x ../data/Phel_genome_index \
    --dta \
    -p 20 \
    -1 ../data/2021_trimmed/$file1 \
    -2 ../data/2021_trimmed/$file2 \
    -S ../analyses/03-hisat2/${base}.sam \
    2> ../analyses/03-hisat2/${base}-hisat.out
done
```


```{bash}
cat ../analyses/03-hisat2/*-hisat.out
```



Convert sam to bam
```{bash}
for samfile in ../analyses/03-hisat2/*.sam; do
  bamfile="${samfile%.sam}.bam"
  sorted_bamfile="${samfile%.sam}.sorted.bam"
  /home/shared/samtools-1.12/samtools view -bS -@ 20 "$samfile" > "$bamfile"
  /home/shared/samtools-1.12/samtools sort -@ 20 "$bamfile" -o "$sorted_bamfile"
  /home/shared/samtools-1.12/samtools index -@ 20 "$sorted_bamfile"
done
```


```{bash}
ls ../analyses/03-hisat2/*sorted.bam | wc -l
```
 

```{r, engine='bash', eval=TRUE}
cat ../analyses/03-hisat2/*hisat.out \
| grep "overall alignment rate"
``` 

# `stringtie`
```{bash}
find ../analyses/03-hisat2/*sorted.bam \
| xargs basename -s .sorted.bam | xargs -I{} \
/home/shared/stringtie-2.2.1.Linux_x86_64/stringtie \
-p 36 \
-eB \
-G ../data/augustus.hints.gtf \
-o ../analyses/03-hisat2/{}.gtf \
../analyses/03-hisat2/{}.sorted.bam
```

steven did this to get the gtf into gff for `stringtie` use
```{bash}
#/home/shared/gffread-0.12.7.Linux_x86_64/gffread \
#../analyses/12-fix-gff/mod_augustus.gtf \
#-T \
#-o ../analyses/13-hisat-deseq2/mod_augustus.gff
```



```{bash}
find ../analyses/03-hisat2/*sorted.bam \
| xargs basename -s .sorted.bam | xargs -I{} \
/home/shared/stringtie-2.2.1.Linux_x86_64/stringtie \
-p 36 \
-eB \
-G ../analyses/13-hisat-deseq2/mod_augustus.gff \
-o ../analyses/03-hisat2/{}.gtf \
../analyses/03-hisat2/{}.sorted.bam
```

```{bash}
eval "$(/opt/anaconda/anaconda3/bin/conda shell.bash hook)"
conda activate
which multiqc


multiqc ../analyses/03-hisat2/ \
-o ../analyses/03-hisat2/
```

# get a count matrix
```{bash}
ls ../analyses/03-hisat2/*gtf
```

copy the above and put into a txt file


```{bash}
cat ../analyses/03-hisat2/list01.txt
```

```{r, engine='bash'}
python /home/shared/stringtie-2.2.1.Linux_x86_64/prepDE.py \
-i ../analyses/03-hisat2/list01.txt \
-g ../data/gene_count_matrix.csv \
-t ../data/transcript_count_matrix.csv
```

```{bash}
head ../data/*matrix.csv
```


# `DESEQ2` for Experiment B

```{r}
library(DESeq2)
library(dplyr)
library(ggplot2)
library(pheatmap)
```

## Differentially expressed transcripts --> Experiment B
read in csv and subset the libraries from experiment B for transcript counts
```{r}
eBtranscriptCounts <- read.csv("../data/transcript_count_matrix.csv", row.names = "transcript_id")
head(eBtranscriptCounts)
```
Subset just the counts for the libraries from Experiment 2:    
```{r}
eBtranscriptCounts <- select(eBtranscriptCounts, PSC.52, PSC.54, PSC.56, PSC.57, PSC.58, PSC.59, PSC.61, PSC.63, PSC.64, PSC.67, PSC.69, PSC.71, PSC.73, PSC.75, PSC.76, PSC.78, PSC.81, PSC.83)
head(eBtranscriptCounts)
```


```{r}
colData <- read.csv("../analyses/03-hisat2/conditions.txt", sep = "\t", row.names = 1)
```

```{r}
eBtranscriptCounts <- as.matrix(eBtranscriptCounts)
```



```{r}
# Check all sample IDs in colData are also in CountData and match their orders
all(rownames(colData) %in% colnames(eBtranscriptCounts)) #should return TRUE
```

```{r}
eBtranscriptCounts <- eBtranscriptCounts[, rownames(colData)]
all(rownames(colData) == colnames(eBtranscriptCounts)) # This should also return TRUE
```

```{r}
# Create a DESeqDataSet from count matrix and labels
dds <- DESeqDataSetFromMatrix(countData = eBtranscriptCounts,
                              colData = colData, design = ~ Condition)


# Run the default analysis for DESeq2 and generate results table
dds <- DESeq(dds)
deseq2.res <- results(dds)

# Sort by adjusted p-value and display
resOrdered <- deseq2.res[order(deseq2.res$padj), ]
vsd <- vst(dds, blind = FALSE)
eBtPCA <- plotPCA(vsd, intgroup = "Condition")
nudge <- position_nudge(y = 4)
eBtPCA + geom_text(aes(label = name), position = nudge)
```

```{r}
# Select top 50 differentially expressed genes
res <- results(dds)
res_ordered <- res[order(res$padj), ]
top_genes <- row.names(res_ordered)[1:50]

# Extract counts and normalize
counts <- counts(dds, normalized = TRUE)
counts_top <- counts[top_genes, ]

# Log-transform counts
log_counts_top <- log2(counts_top + 1)

# Generate heatmap
pheatmap(log_counts_top, scale = "row")

```

```{r}
# Count number of hits with adjusted p-value less then 0.05
dim(res[!is.na(deseq2.res$padj) & deseq2.res$padj <= 0.05, ])
```




all controls are in same clump except 56, no clue why.
But, since the repeat control star samples are in the same clump, I guess it doesn't really matter which ones I remove to balance the analyses. 

also.... for exposed, two stars were sampled pre arm drop. 59 was at arm twist on day 9, and 83 was at arm cross on day 13. so maybe i'll drop those two as well and just do 6 arm drop vs 6 control.

here are the n=6 control vs n=6 exposed libraries i'll continue with:

| library_ID | treatment_group | experiment_day | disease_sign |
|------------|-----------------|----------------|--------------|
| PSC-52     | control         | 9              | NA           |
| PSC-54     | control         | 9              | NA           |
| PSC-57     | exposed         | 9              | armdrop      |
| PSC-61     | control         | 10             | NA           |
| PSC-64     | control         | 10             | NA           |
| PSC-67     | exposed         | 10             | armdrop      |
| PSC-69     | exposed         | 10             | armdrop      |
| PSC-71     | exposed         | 10             | armdrop      |
| PSC-73     | control         | 11             | NA           |
| PSC-75     | exposed         | 11             | armdrop      |
| PSC-76     | control         | 12             | NA           |
| PSC-78     | exposed         | 12             | armdrop      |








