---
title: "15-hisat2-deseq2-summer2022"
output: html_document
date: "2024-09-02"
---
Follow code from 03-hisat2-summer2021_2022.Rmd. A lot of steps were done in the beginning that don't need to be repeated. 

Summer 2022    
`/home/shared/8TB_HDD_02/graceac9/data/pycno2022` 

```{r, engine='bash'}
cd ../data

```



```{r, engine='bash', eval=TRUE}
head ../data/augustus.hints.gtf

```

For `stringtie` later, the gtf needs to be in gff format. There isn't currently one in existence that works. Steven figured out how to fix this issue (https://sr320.github.io/tumbling-oysters/posts/sr320-20-seastar/)  

Won't run code because Steven did this already and the new file lives in:
` ../analyses/12-fix-gff/mod_augustus.gtf`

```{r}
library(knitr)
library(tidyverse)
library(Biostrings)
library(pheatmap)
library(DESeq2)
library(tidyverse)
```

# 1. Build an index
Follow this code: https://htmlpreview.github.io/?https://github.com/urol-e5/deep-dive/blob/8fd4ad4546d1d95464952f0509406efd9e42ffa0/D-Apul/code/04-Apulcra-hisat.html 

```{bash}
pwd
```

From the gtf, get exon list
```{bash}
/home/shared/hisat2-2.2.1/hisat2_extract_exons.py \
../analyses/12-fix-gff/mod_augustus.gtf \
> ../analyses/15-hisat2-deseq2-summer2022/m_exon.tab
```


```{bash}
head ../analyses/15-hisat2-deseq2-summer2022/m_exon.tab
```

from the gtf, get the splice sites
```{bash}
#!/bin/bash

# This script will extract splice sites from the gtf file

# This is the command to extract splice sites from the gtf file
/home/shared/hisat2-2.2.1/hisat2_extract_splice_sites.py \
../analyses/12-fix-gff/mod_augustus.gtf \
> ../analyses/15-hisat2-deseq2-summer2022/m_splice_sites.tab

```

use the genome fasta to make an index for alignment
```{bash}
# build an index 
/home/shared/hisat2-2.2.1/hisat2-build \
../data/ncbi_dataset/data/GCA_032158295.1/GCA_032158295.1_ASM3215829v1_genomic.fna \
../data/Phel_genome_index \
--exon ../analyses/15-hisat2-deseq2-summer2022/m_exon.tab \
--ss ../analyses/15-hisat2-deseq2-summer2022/m_splice_sites.tab \
-p 40 \
../analyses/12-fix-gff/mod_augustus.gtf \
2> ../analyses/15-hisat2-deseq2-summer2022/hisat2-build_stats.txt
```


```{bash}
tail ../analyses/03-hisat2/hisat2-build_stats.txt
```


# 2. ALIGNMENT
Trimmed data in `/home/shared/8TB-`/home/shared/8TB_HDD_02/graceac9/data/pycno2022`
```{bash}
pwd
```



```{bash}
for file in ../../../data/pycno2022/*_R1*fq.gz; do
    # Remove the  part to get the base name
    base=$(basename "$file" _R1_001.fastq.gz.fastp-trim.20231101.fq.gz)
    # Construct the names of the pair of files
    file1=${base}_R1_001.fastq.gz.fastp-trim.20231101.fq.gz
    file2=${base}_R2_001.fastq.gz.fastp-trim.20231101.fq.gz
    # Run the hisat2 command
    /home/shared/hisat2-2.2.1/hisat2 \
    -x ../data/Phel_genome_index \
    --dta \
    -p 20 \
    -1 ../../../data/pycno2022/$file1 \
    -2 ../../../data/pycno2022/$file2 \
    -S ../analyses/15-hisat2-deseq2-summer2022/${base}.sam \
    2> ../analyses/15-hisat2-deseq2-summer2022/${base}-hisat.out
done
```

```{bash}
cat ../analyses/15-hisat2-deseq2-summer2022/*-hisat.out
```

Convert sam to bam
```{bash}
for samfile in ../analyses/15-hisat2-deseq2-summer2022/*.sam; do
  bamfile="${samfile%.sam}.bam"
  sorted_bamfile="${samfile%.sam}.sorted.bam"
  /home/shared/samtools-1.12/samtools view -bS -@ 20 "$samfile" > "$bamfile"
  /home/shared/samtools-1.12/samtools sort -@ 20 "$bamfile" -o "$sorted_bamfile"
  /home/shared/samtools-1.12/samtools index -@ 20 "$sorted_bamfile"
done
```

```{bash}
ls ../analyses/15-hisat2-deseq2-summer2022/*sorted.bam | wc -l
```

```{r, engine='bash', eval=TRUE}
cat ../analyses/15-hisat2-deseq2-summer2022/*hisat.out \
| grep "overall alignment rate"
``` 

# `stringtie`
```{bash}
find ../analyses/15-hisat2-deseq2-summer2022/*sorted.bam \
| xargs basename -s .sorted.bam | xargs -I{} \
/home/shared/stringtie-2.2.1.Linux_x86_64/stringtie \
-p 36 \
-eB \
-G ../analyses/13-hisat-deseq2/mod_augustus.gff \
-o ../analyses/15-hisat2-deseq2-summer2022/{}.gtf \
../analyses/15-hisat2-deseq2-summer2022/{}.sorted.bam
```

```{bash}
eval "$(/opt/anaconda/anaconda3/bin/conda shell.bash hook)"
conda activate
which multiqc


multiqc ../analyses/15-hisat2-deseq2-summer2022/ \
-o ../analyses/15-hisat2-deseq2-summer2022/
```

# get a count matrix
```{bash}
ls ../analyses/15-hisat2-deseq2-summer2022/*gtf
```

copy the above and put into a txt file

write library name to left of path

```{bash}
cat ../analyses/15-hisat2-deseq2-summer2022/list01.txt
```

```{r, engine='bash'}
python /home/shared/stringtie-2.2.1.Linux_x86_64/prepDE.py \
-i ../analyses/15-hisat2-deseq2-summer2022/list01.txt \
-g ../data/gene_count_matrix_2022.csv \
-t ../data/transcript_count_matrix_2022.csv
```

```{bash}
head ../data/*matrix_2022.csv
```

# `DESEQ2` for Experiment C

```{r}
library(DESeq2)
library(dplyr)
library(ggplot2)
library(pheatmap)
library("pcaExplorer")
library("airway")
```


Install 2024-09-06 - `pcaExplorer`:
```{r}
#if (!require("BiocManager", quietly = TRUE))
    install.packages("BiocManager")

#BiocManager::install("pcaExplorer")
```
```{r}
#if (!requireNamespace("BiocManager", quietly=TRUE))
    install.packages("BiocManager")
#BiocManager::install("airway")
```



## Explore all the samples with `pcaExplorer` 
read in csv 
```{r}
eCtranscriptCounts <- read.csv("../data/transcript_count_matrix_2022.csv", row.names = "transcript_id")
head(eCtranscriptCounts)
```

made a coldata frame in google sheets https://docs.google.com/spreadsheets/d/1tZuBu8wFnKgh_mvZY8jpleF4nzMsQh-NOHZ1YKoeyqU/edit?gid=0#gid=0 

downloaded as .csv, then copy-pasted into data directory and pushed to github 

it's call coldata2022.csv in data in this repo

```{r}
pcaExplorer()
```




Subset just the counts for the libraries of interest

Some comparisons that can be made:

Exposed vs Exposed with age class being the treatment --> 5 small vs 5 large:    

Make a PCA of all samples:
```{r}

```




